tree-sitter/src/TreeSitter/Unmarshal/Examples.hs

```diff
@@ -1,62 +1,93 @@
 {-# LANGUAGE DeriveAnyClass, DeriveGeneric, DuplicateRecordFields, TypeOperators #-}
 module TreeSitter.Unmarshal.Examples () where
 
+import Control.Effect.Reader
+import Control.Monad.Fail
+import qualified Data.ByteString as B
 import qualified Data.Text as Text
-import GHC.Generics ((:+:), Generic)
+import qualified Data.Text.Encoding as Text
+import GHC.Generics ((:+:), Generic1)
+import Numeric (readDec)
+import Prelude hiding (fail)
+import Source.Range
 import TreeSitter.Unmarshal
 
 -- | An example of a sum-of-products datatype.
 data Expr a
   = IfExpr (If a)
   | BlockExpr (Block a)
+  | VarExpr (Var a)
   | LitExpr (Lit a)
   | BinExpr (Bin a)
-  deriving (Generic, Unmarshal)
+  deriving (Generic1, Unmarshal)
 
 -- | Product with multiple fields.
 data If a = If { ann :: a, condition :: Expr a, consequence :: Expr a, alternative :: Maybe (Expr a) }
-  deriving (Generic, Unmarshal)
+  deriving (Generic1, Unmarshal)
 
-instance SymbolMatching (If a) where
+instance SymbolMatching If where
   symbolMatch _ _ = False
   showFailure _ _ = ""
 
 -- | Single-field product.
 data Block a = Block { ann :: a, body :: [Expr a] }
-  deriving (Generic, Unmarshal)
+  deriving (Generic1, Unmarshal)
 
-instance SymbolMatching (Block a) where
+instance SymbolMatching Block where
   symbolMatch _ _ = False
   showFailure _ _ = ""
 
 -- | Leaf node.
-data Lit a = Lit { ann :: a, text :: Text.Text }
-  deriving (Generic, Unmarshal)
+data Var a = Var { ann :: a, text :: Text.Text }
+  deriving (Generic1, Unmarshal)
 
-instance SymbolMatching (Lit a) where
+instance SymbolMatching Var where
+  symbolMatch _ _ = False
+  showFailure _ _ = ""
+
+-- | Custom leaf node.
+data Lit a = Lit { ann :: a, lit :: IntegerLit }
+  deriving (Generic1, Unmarshal)
```

Comment:
I wanted to illustrate something cool the `Unmarshal` machinery can do: it can accommodate datatypes which a) our TH code doesn’t generate, and b) do something more specialized. See the `UnmarshalAnn` instance for `IntegerLit`, below.